### PR TITLE
Parsing Collections, Environments, and Non-Blocking Async OH MY

### PR DESCRIPTION
**Pull request by mblydenburgh**
- add new menu bar with buttons to import/export/create things
- add parsing logic for collection and environment files
- create simple import modal that calls back end to start import
- parse file and print message for now (can call db save method with this parsed struct?)
- add support for async main function in GUI. Needed to make accessed data shared between threads
- perform async operations in a non-blocking way

Ideally I was thinking we could support a file drag/drop to get the file path but starting simple with a text field that you have to copy/paste the path in to read. I think though that collections and environments would be ready for use in to make db models?

### CODE REVIEW

File: api/src/domain/collection.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct CollectionRequestHeader {
+    pub key: String,
+    pub value: String,
+    pub r#type: String,
+}
+
+// Custom impl for Deserialize to determine wheth parsed thing is a request or sub folder
+impl<'de> Deserialize<'de> for CollectionItemOrFolder {
```
Comment (mblydenburgh) [inline api/src/domain/collection.rs:97]: Personally, I like how standardized Rust's Trait ecosystem make things. When build functionality you can always have a common abstraction for behavior, especially being so core to language.

File: api/tests/api/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+use api::PostieApi;
+
+pub struct TestApp {
```
Comment (mblydenburgh) [inline api/tests/api/helpers.rs:3]: I kind of based this testing pattern around what was introduced in Zero2Prod - having a main TestApp that has some reusable helper functions and have a place to stand up other test clients (like db), and perform any other initialization.

File: gui/src/main.rs [hunk 3]
```diff
@@ -56,32 +70,45 @@ impl Default for GuiConfig {
...
-    fn spawn_submit(&mut self, input: HttpRequest) -> Result<Value, Box<dyn Error>> {
-        let result = self.rt.block_on(async {
-            let api_response = PostieApi::make_request(input).await;
-            api_response
+    async fn submit(input: HttpRequest) -> Result<Value, Box<dyn Error>> {
+        PostieApi::make_request(input).await
+    }
+    fn spawn_submit(&self, input: HttpRequest) -> Result<(), Box<dyn Error>> {
+        let result_for_worker = self.response.clone();
+        tokio::spawn(async move {
```
Comment (mblydenburgh) [inline gui/src/main.rs:98]: Set up a synchronous function that spins up a non-blocking thread that then calls the async function.

File: gui/src/main.rs [hunk 4]
```diff
@@ -56,32 +70,45 @@ impl Default for GuiConfig {
...
 impl Gui {
-    fn spawn_submit(&mut self, input: HttpRequest) -> Result<Value, Box<dyn Error>> {
-        let result = self.rt.block_on(async {
-            let api_response = PostieApi::make_request(input).await;
-            api_response
+    async fn submit(input: HttpRequest) -> Result<Value, Box<dyn Error>> {
+        PostieApi::make_request(input).await
+    }
+    fn spawn_submit(&self, input: HttpRequest) -> Result<(), Box<dyn Error>> {
+        let result_for_worker = self.response.clone();
```
Comment (mblydenburgh) [inline gui/src/main.rs:97]: Creating a clone of the Mutex that holds the pointer in memory for `response` that the new thread is going to request a lock on.

File: gui/src/main.rs [hunk 5]
```diff
@@ -55,33 +62,44 @@ impl Default for GuiConfig {
...
+        tokio::spawn(async move {
+            match Gui::submit(input).await {
+                Ok(res) => {
+                    println!("Res: {}", res);
+                    let mut result_write_guard = result_for_worker.try_write().unwrap();
+                    *result_write_guard = Some(res);
```
Comment (mblydenburgh) [inline gui/src/main.rs:95]: From inside the other thread, the * is the dereference operator. This is essentially updating the value of the memory in the main thread.

File: gui/src/main.rs [hunk 6]
```diff
@@ -55,33 +62,44 @@ impl Default for GuiConfig {
...
+    async fn submit(input: HttpRequest) -> Result<Value, Box<dyn Error>> {
+        PostieApi::make_request(input).await
+    }
+    fn spawn_submit(&self, input: HttpRequest) -> Result<(), Box<dyn Error>> {
+        let result_for_worker = self.response.clone();
+        tokio::spawn(async move {
+            match Gui::submit(input).await {
+                Ok(res) => {
+                    println!("Res: {}", res);
+                    let mut result_write_guard = result_for_worker.try_write().unwrap();
```
Comment (mblydenburgh) [inline gui/src/main.rs:94]: .try_write() attempts to gain write access from the RwLock

File: gui/src/main.rs [hunk 7]
```diff
@@ -55,33 +62,44 @@ impl Default for GuiConfig {
...
+            import_file_path: String::from(""),
+            import_mode: ImportMode::COLLECTION,
         }
     }
 }
 
 pub struct Gui {
-    pub config: GuiConfig,
-    pub rt: runtime::Runtime,
+    pub config: Arc<RwLock<GuiConfig>>,
```
Comment (mblydenburgh) [inline gui/src/main.rs:73]: Arc is atomically reference counted, essentially letting multiple threads have access to this memory.

RwLock is a Read-Write Lock. It lets multiples difference owners have read access, but only one can have write access at a time.

File: gui/src/main.rs [hunk 8]
```diff
@@ -97,195 +115,237 @@ impl Gui {
...
-            ActiveWindow::ENVIRONMENT => {
-                ui.label("Environments");
-            }
-            ActiveWindow::HISTORY => {
-                ui.label("History");
-            }
-        });
-        TopBottomPanel::top("top_panel").show(ctx, |ui| {
-            ui.heading("Welcome to Postie!");
+    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
```
Comment (mblydenburgh) [inline gui/src/main.rs:118]: cargo fmt made a lot of formatting changes.

File: gui/src/main.rs [hunk 9]
```diff
@@ -97,195 +115,237 @@ impl Gui {
...
+                        };
+                    });
+                });
             });
-            ui.horizontal(|ui| {
-                if ui.button("Params").clicked() {
-                    self.config.request_window_mode = RequestWindowMode::PARAMS;
+        });
+        SidePanel::left("nav_panel").show(ctx, |ui| {
+            if let Ok(mut config) = self.config.try_write() {
```
Comment (mblydenburgh) [inline gui/src/main.rs:156]: Any bits of ui that go and update the `config` property are wrapped in this if let statement for acquiring a write lock.

File: gui/src/main.rs [hunk 10]
```diff
@@ -97,195 +115,237 @@ impl Gui {
...
+                                ui.add(
+                                    TextEdit::multiline(&mut config.body_str)
+                                        .code_editor()
+                                        .desired_rows(20)
+                                        .lock_focus(true)
+                                        .font(TextStyle::Monospace),
+                                );
+                            });
+                        });
+                    if self.response.try_read().unwrap().is_some() {
```
Comment (mblydenburgh) [inline gui/src/main.rs:254]: If something doesn't need to write, it can just request read access though.

**Review by tqhoughton — APPROVED** [review]
I understand all of this and none of it went over my head

**Comment by tqhoughton** [comment]
Also I think current `main` is broken when you attempt to send a request but I checked out this branch and it looks like it fixed whatever that issue was

**Comment by mblydenburgh** [comment]
> Also I think current `main` is broken when you attempt to send a request but I checked out this branch and it looks like it fixed whatever that issue was

Ah yea, that was due to the addition of [tokio::main] to the main function. That is a macro that, at compile time, injects code. This particular one essentially does wraps the main function in the following code:
```
fn main() {
  tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .unwrap()
        .block_on(async {
            ....
        })
}
```

Since the Gui struct was already creating a tokio runtime inside the code and running its own .block_on, that was causing two things to try to use the main thread to run blocking code which caused the error.  If you take the current `main`, remove the tokio macro and make `main` no async, things will work again. You'll obv have to comment out the db code too.

**Pull request merged**